odoo/addons/base/models/res_users.py

```diff
@@ -498,7 +498,8 @@ def onchange(self, values, field_names, fields_spec):
         # Hacky fix to access fields in `SELF_READABLE_FIELDS` in the onchange logic.
         # Put field values in the cache.
         if self == self.env.user:
-            [self.sudo()[field_name] for field_name in self.SELF_READABLE_FIELDS]
+            readable_fields = frozenset(self.SELF_READABLE_FIELDS)
+            [self.sudo()[field_name] for field_name in readable_fields]
```

Comment:
By the way, I have the feeling that this hack is no longer necessary, since fetching data in cache _no longer_ gives access to those fields.

Reply:
I think it is still necessary because of the `sudo`. We just add a check here that the field is accessible, but the record is still not readable by the user.

Reply:
Oh, I get it now!  That's for computing the initial snapshot in `onchange`, which is done this way:
- fetch the origin record in cache;
- copy those values to the new record's cache;
- adapt the new record's cache with the values coming from the caller;
- create the snapshot object based on the new record's cache values.

Fetching the private fields as `sudo()` enables the first step to work as expected :+1:

Reply:
By the way, why using a `frozenset`?  I don't see the point.  I don't think there are _that many_ duplicates in `self.SELF_READABLE_FIELDS`.  And memoizing `self.sudo()` would probably be more efficient.  What about this instead?
```suggestion
            self.sudo().fetch(self.SELF_READABLE_FIELDS)
```